contrib/dracut/90zfs/zfs-load-key.sh.in

```diff
@@ -0,0 +1,56 @@
+#!/bin/bash
+
+# This script only gets executed on systemd systems, see mount-zfs.sh for non-systemd systems
+
+# There is a race between the zpool import and the pre-mount hooks, so we
+#   we wait for a pool to be imported
+while true; do
+    if $(zpool list -H | grep -q -v '^$'); then
+        break
+    fi
+    if $(systemctl is-failed zfs-import-scan.service) == 'failed' ]]; then
+      exit 1
+    fi
+    if [[ $(systemctl is-failed zfs-import-cache.service) == 'failed' ]]; then
+      exit 1
+    fi
+    sleep 0.10s
```

Comment:
What I do not understand is why this script must execute as a Dracut pre-mount hook.  This ain't right.

In non-systemd scenarios, this script must execute after pool import.

In systemd scenarios, this script must execute (without loops or retries) exactly once after both zpool-import-*.service.

I see nothing in the generator that wires this up sandwiched between the zpool-import-* units and before sysroot.mount unit.

Please fix this.

Reply:
from what I could tell, the hook only executes on systemd systems (otherwise mount-zfs.sh is called).  What seems to happen in testing on systemd systems is that it executes in parallel with sysroot.mount (so the dataset is not unlocked in time).  This was verified with debug prints.